### ISSUE TITLE
Enable dynamic command names

### ISSUE THREAD

**Issue by Nikey646** [Nihlus/Remora.Commands#2]
Being capable of defining the group or command name at runtime would be nice.

This would allow applications to use an external source (EG: a configuration file) to define the structure of commands.

**Comment by Nihlus** [comment]
It'd be an interesting thing to have - I am, however, not too sure how I'd make an API like that look. Do you have any examples of how you'd like to use it, or how you'd hook into the various parameters of the defined commands?

Spitballing, I think my first go at it would start by abstracting out information about the structure of a command into a separate container, and having that container be mutable at group registration time (probably via some fluent API, allowing runtime modification/overriding of fallback attributes).

Maybe something like this?

```cs
serviceCollection.AddCommandGroup<MyGroup>
(
    g => g
        .WithCommand(MyGroup.MyCommand)
        .WithNamedParameter
        ( 
            // unsure of how to access parameters here - by name, index, or either/both? //
            longName: "my-parameter",
            shortName: 'p'
        )
)
```

I know Discord.NET has something like this, but I'm not familiar with that portion of their API.

**Comment by Nikey646** [comment]
The start of the Discord.Net implementation is [`CommandService.CreateModuleAsync`](https://github.com/discord-net/Discord.Net/blob/084db253f32269e364c7106898f2c1b34a8727ff/src/Discord.Net.Commands/CommandService.cs#L136)

The usage for the metadata that Remora.Commands has similar to 
```cs
// ModuleBuilder is the type
commandService.CreateModuleAsync(group, (ModuleBuilder b) => b.AddCommand(command, ExecuteCommandFunc, CreateCommandFunc);

void CreateCommandFunc(CommandBuilder builder)
{
    builder.AddAliases("b", "c")
                .AddParameter(name, type, CreateParameterFunc);
}

void CreateParameterFunc(ParameterBuilder builder) 
{
    builder.TypeReader = myTypeReader;
}
```

A similar pattern could be used if you wanted to do something similar to Discord.Net.

**Comment by Nihlus** [comment]
Right, I see, so it creates groups without involving declaring a class at all. That's interesting - it's a lot more involved than I first considered.

**Comment by Nikey646** [comment]
Correct, it basically allows the user to construct the equivalent of the CommandNode / GroupNode via an API.
